pkg/bpf/bpf_linux.go

```diff
@@ -113,7 +113,16 @@ func UpdateElementFromPointers(fd int, structPtr unsafe.Pointer, sizeOfStruct ui
 	}
 
 	if ret != 0 || err != 0 {
-		return fmt.Errorf("Unable to update element for map with file descriptor %d: %s", fd, err)
+		switch err {
+		case unix.E2BIG:
+			return fmt.Errorf("Unable to update element for %s map with file descriptor %d: the map is full, please consider resizing it. %w", mapName, fd, err)
```

Comment:
I feel a bit uneasy about the integration of dynamic content into the error message here, something seems a bit broken with the abstraction since I would otherwise expect the actual log message to include the map name as a dedicated field in structured logging, ie:

```
log.WithFields(logrus.Fields{
    logfields.MapName: "lb",
    ...
}.WithError(err).Warning("Cannot update service map")
```

So I feel like if we actually studied the various locations where these error messages end up getting logged, we could probably slim down the actual error here to the crucial details. The goal of the PR being to introduce the "the map is full" / "specified key already exists" / "key does not exist" information, which the underlying errno messages do not properly convey to the user.

In my mind, the "Unable to update element" bit should be obvious from the higher layer logging message, but for safety that part seems fine to keep here.

The map name I'd expect to be properly handled by the higher layers; if you have example error messages where this is unclear then I'd be curious to see them and understand why we can't improve those error messages closer to the `log....` call rather than deep in the stack.

While we're reviewing these messages, I wonder whether we've used the file descriptor info before or if there's a reasonable way to actually use this information. File descriptor is just a number, if it were invalid then the kernel will give us back `EBADFD`. But I don't think there's any way to figure out which map the file descriptor corresponds to so does it really provide any more information that could be useful while debugging an issue? Maybe we can just drop it.

One other nit, wrapping error messages usually is done with a colon before the wrap format bit rather than full stop, ie please use `consider resizing it: %w` rather than `consider resizing it. %w`. Same for the other logs below.

Reply:
> > Beyond that I have a bit of unease about the way that these error messages are integrating dynamic content, thereby moving away from structured logging. We've almost completely switched Cilium over to structured logs to help logging systems do a better job of indexing and aggregating logs based on the field types of the logs. Any move away from that approach feels like it introduces more work for us to improve it again in future. Structured logging also helps users by clearly separating different details of the message into fields like `level=error msg="Cannot update service map" map="lb" error="unable to update element: map is full" hint="You can change the size of the LB map with the flag "bpf-lb-map-max". Resizing the map might break existing connections to services."`

> You lost me. I output the error message following the way they were done in the code before me. Are you suggesting to completely change it in this PR?

The error message before included the fd, and I think that's mainly because we haven't revisited this code since we started pushing more on structured logging. In my proposal above, I even suggest maybe we shouldn't be logging the fd at all.

This PR currently proposes to take this even further by propagating the map names down through the layers to embed into the errors here. That's the piece that seems not quite right to me. I think that it's a good idea to have the map name in the logs, but I think we should do it in a structured manner, ie log the map name directly from the log statements rather than passing it down to this layer and embedding the map name in the error.

What I anticipate is that if we accept this particular aspect as-is in the PR, then in future we still need to remove the map name from the error again so that the dynamic content is properly structured for better integration with loggers (whether it's just someone grepping through the logs to find similar logs with the same `mapName` or full on log processing engines that correlate logs together based on the fields we log).

Looking back over the review, it seems I missed the original time that this was brought up (https://github.com/cilium/cilium/pull/16034#discussion_r642545651) cc @brb .